### ISSUE TITLE
Webpacker pack from S3+Cloudfront on Heroku does not take the manifest from Cloudfront and the config.action_controller.asset_host value

### ISSUE THREAD

**Issue by thebravoman** [rails/webpacker#2619]
The public/packs are sync to an s3 bucket, same as the public/assets. 

Delivering assets from public/assets works.

But when trying to open anything dependent on public/packs I get.
```
2020-06-04T05:33:03.622234+00:00 app[web.1]: [58f2d221-761b-44ec-807c-6b922342c10b] method=GET path=/teach/organizations format=html controller=OrganizationsController action=index status=500 error='ActionView::Template::Error: Webpacker can't find admin_unify_style.css in /app/public/packs/manifest.json. Possible causes:
2020-06-04T05:33:03.622242+00:00 app[web.1]: 1. You want to set webpacker.yml value of compile to true for your environment
2020-06-04T05:33:03.622242+00:00 app[web.1]: unless you are using the `webpack -w` or the webpack-dev-server.
2020-06-04T05:33:03.622243+00:00 app[web.1]: 2. webpack has not yet re-run to reflect updates.
2020-06-04T05:33:03.622243+00:00 app[web.1]: 3. You have misconfigured Webpacker's config/webpacker.yml file.
2020-06-04T05:33:03.622244+00:00 app[web.1]: 4. Your webpack configuration is not creating a manifest.
2020-06-04T05:33:03.622244+00:00 app[web.1]: Your manifest contains:
2020-06-04T05:33:03.622245+00:00 app[web.1]: {
2020-06-04T05:33:03.622245+00:00 app[web.1]: }
```

The manifest contains 
```
"entrypoints": {

    "admin_unify_style": {
        "css": [
            "/packs/css/admin_unify_style-c2ad729a.css"
        ],
        "js": [
            "/packs/js/admin_unify_style-62270c86710162620f91.js"
        ],
        "js.map": [
            "/packs/js/admin_unify_style-62270c86710162620f91.js.map"
        ]
```
but it is on S3, it is not available in the app filesystem on heroku
I have 

```
  config.action_controller.asset_host = "https://....cloudfront.net"

```
It is working for /assets

I think there should be a way for webpacker to find that the packs are on s3 and to request them from there and not in a app.  We should not be asked to make
```
rake assets:precompile 
```
on heroku. The assets are already precompiled and synced to S3. 

The asset itself is available on cloudfront. https://d2nmr6p48f8xwg.cloudfront.net/packs/css/admin_unify_style-c2ad729a.css. It is just that webpacker expects the manifest to be on heroku and not on cloudfront. 

I looked through the deployment page and especially https://github.com/rails/webpacker/blob/master/docs/deployment.md#cdn where it says 
```
Webpacker out-of-the-box provides CDN support using your Rails app config.action_controller.asset_host setting. If you already have CDN added in your Rails app you don't need to do anything extra for Webpacker, it just works.
```
I have the heroku buildpacks
```
2. heroku/ruby
3. heroku/nodejs
```

Update 3:
There are also not devDependencies in package.json except
```
  "devDependencies": {
    "webpack-dev-server": "^3.11.0"
  },
```

**Comment by thebravoman** [comment]
I have tried setting WEBPACKER_ASSET_HOST - because this is used in the code -  https://github.com/rails/webpacker/search?q=WEBPACKER_ASSET_HOST&unscoped_q=WEBPACKER_ASSET_HOST

For future reference - it did not work. Still can not get the manifest from the CDN

I is also very interesting how the documentation mentions Webpacker uses config.action_controller.asset_host but when you search for asset_host in the code and how it is used you get only the mention in the documentation - https://github.com/rails/webpacker/search?q=asset_host&unscoped_q=asset_host

**Comment by rossta** [comment]
A few pieces of info based on my understanding:

Webpacker expects to find the manifest.json file on disk, according to https://github.com/rails/webpacker/blob/9d15b93edab5c53f0510dab309fedcece54e612f/lib/webpacker/manifest.rb#L84. So the typical use case would expect the webpack manifest.json must exist within the context of where the application server is running, just like the requirement for .sprockets-manifest.json. This would explain why Rails would not find it even with asset host settings; these only apply to the public static assets.

Since the manifest.json file must be kept up-to-date with any changes to assets and itself does not get fingerprinted (i.e., the location is always `public/packs/manifest.json`), it wouldn't be typical to read from CDN where assets would typically respond with long-lived HTTP caching headers. 

How do you make the Sprockets manifest available if not built through Heroku?

**Comment by thebravoman** [comment]
Here is the summary for now.
1. I should have public/assets/.sprockets-manifest because of sprockets. So I have it but all the assets are on s3 from where cloudfront is getting them. That works.
2. But webpacker comes a long and sees the public/assets and thinks - "There is a manifest I am not compilting anything". It took me hours to realize this. So it is not compiling anything even on heroku.
3. I just had to commit 33MB of assets to my repo to make at least the styles work, which is arguably like and ftp uploads that once were. 

4. I would now try to commit only the manifest file of the webpacker and leave all the other files precompiled on S3 from where cloudfront is taking them.

**Comment by thebravoman** [comment]
Looking at today's experience the most unexpected thing for me was that webpacker on heroku was seeing that there is a public/assets/.sprockets* and was assuming this means compiled public/packs 

```
remote:        Detected manifest file, assuming assets were compiled locally        
``` 

But the error message after that suggest that something was wrong with the manifest. Since the manifest does not have a has and since the webpacker documentation for CDN says - "just add asset_host and it just works" it took me some time to realize that webpacker was looking at another manifest. This was rather quick, but with 2 tickets with of heroku itself made it kind of difficult.

Probably a section of how to keep the public/packs out of the repo - because it is large and mainly vendor is something that could be helpful.

**Comment by cmacduff** [comment]
I'm curious about this same setup. How has committing the manifests worked out for you? Seems like it'd make PR's hard to manage with the constant changes to manifest.json.

I'm using https://github.com/AssetSync/asset_sync to get the files to S3 if that makes any difference (doesn't seem too relevant). @thebravoman, you might be able to remove the s3cmd with this gem? I got it to upload the fonts just fine.

The stack is basically the same except Hatchbox instead of Heroku, but I doubt that will change implementation that much.

I'd be very curious to know if webpacker supports a streamlined setup for having assets/packs in S3. Part of the orchestration problem I see is `manifest.json` isn't unique. It seems like it needs a digest of some sort in the file name so deployments don't have to worry about newly sync'd code to the S3 bucket overwriting the manifest.json.

**Comment by thebravoman** [comment]
Looking back at this I think I found a way to make webpacker works for us and I don't consider this an issue. Probably it should be closed. It was mostly my misunderstanding along with things I did not know and could not found in the documentation.

@cmacduff the process has made a couple of thousand builds to production for us and it works. We commit the manifest and sync the assets with s3.

Thanks for the suggestion for asset_sync. What a decade of running rails platforms has taught me is that you start with a gem that wraps the underlying tool but you slowly migrate to using the underlying tool itself and I like having the s3cmd.
